corehq/apps/api/validation.py

```diff
@@ -11,16 +20,64 @@
     TableauRoleValidator,
     CustomDataValidator,
     EmailValidator,
+    UserAccessValidator,
+    UserRetrievalResult,
 )
 from corehq.apps.users.validation import validate_primary_location_assignment
 from corehq.apps.registration.validation import AdminInvitesUserFormValidator
+from corehq.toggles import TABLEAU_USER_SYNCING
+
+
+@dataclass
```

Comment:
This `dataclass` plus custom validator class makes me think of `attrs`, which has a [built-in system for writing custom validation](https://www.attrs.org/en/stable/init.html#validators).  That'd I think cut down on some of the boilerplate, and reduce the mental burden of understanding a new system.

I do also see that tastypie lets you [pass in a Django form](https://django-tastypie.readthedocs.io/en/latest/validation.html#cleaneddataformvalidation) for validation.  That'd be cool if we could reuse the same form class for the UI, the API, and the importer, but I'm guessing there'd be some blockers.

Reply:
I'll look into `attrs`

It would be nice to use the same django form for validation, but would require some overhauling. I already did some refactoring to consolidate the validation between the UI and importer in a [previous PR](https://github.com/dimagi/commcare-hq/pull/35448) and tried in this PR to reuse importer validation logic for the API.

Reply:
Thanks for that suggestion - it overall is a bit nicer. I implemented it in the follow up "Update Web User PR" to avoid it conflicting with the files in this PR https://github.com/dimagi/commcare-hq/pull/35617/commits/25b7fe21458b9bf4b3b299e0fcfb722e2ea8da1f